functionary/prompt_template/llama3_prompt_template.py

```diff
@@ -0,0 +1,301 @@
+        text_content = None
+        tool_call_start_index = 0
+
+        if not llm_output.startswith(
+            self.function_separator
+        ):  # if first token is not function_call
+            text_content = chunks[0]
+            tool_call_start_index = 1
+
+        for chunk in chunks[tool_call_start_index:]:
+            # format: function_name\narguments<end_of_functioncall>
+            index = chunk.find("\n")
+            func_name = chunk[:index].strip()
+            arguments = chunk[index + 1 :].strip()
+            tool_calls.append(
+                {
+                    "function": {"name": func_name, "arguments": arguments},
+                    "id": prompt_utils.get_random_tool_call_id(),
+                    "type": "function",
+                }
+            )
+        return {"role": "assistant", "content": text_content, "tool_calls": tool_calls}
+
+    def convert_message_to_prompt(self, message: Dict) -> str:
+        role = message["role"]
+        content = message.get("content", None)
+        if role == "tool":
+            tool_name = message["name"]
+            content = f"name={tool_name}\n{content}"
+
+        prompt_template = (
+            "<|start_header_id|>%s<|end_header_id|>\n\n{text}<|eot_id|>" % role
+        )
+
+        if role in ["user", "system", "tool"]:
+            return prompt_template.format(text=content)
+
+        assert role == "assistant"
+        tool_calls = message.get("tool_calls", [])
+        if tool_calls is None:
+            tool_calls = []
+
+        if content is None and len(tool_calls) == 0:
+            return f"<|start_header_id|>{role}<|end_header_id|>\n\n"
+
+        if content is not None and len(tool_calls) == 0:  # text-only
+            return prompt_template.format(text=content)
+
+        tool_call_prompts = []
+        for tool_call in tool_calls:
+            arguments = tool_call["function"]["arguments"]
+            tool_name = tool_call["function"]["name"]
+            tool_prompt = f"{tool_name}\n{arguments}"
+            tool_call_prompts.append(tool_prompt)
+
+        if (
+            content is None and len(tool_calls) > 0
+        ):  # function call only (<sep>fc1<sep>fc2<sep>)
+            tool_call_content = self.function_separator + self.function_separator.join(
+                tool_call_prompts
+            )
+            return prompt_template.format(text=tool_call_content)
+
+        # Here is the case contains both text-response and tool_calls (content<sep>fc1<sep>fc2<sep>)
+        total_content = (
+            content
+            + self.function_separator
+            + self.function_separator.join(tool_call_prompts)
+        )
+        return prompt_template.format(text=total_content)
+
+    def pre_process_messages_before_inference(self, messages: List[Dict]) -> List[Dict]:
+        return prompt_utils.reorder_tool_messages_by_tool_call_ids(messages)
+
+    def update_state_for_function(self, current_state):
```

Comment:
Maybe this can be a private method instead?

Reply:
This should also have docstring

Reply:
Sure I will add docstring